File: MMR Tracker V3/TrackerObjects/CheckableLocation.cs
﻿using static MMR_Tracker_V3.TrackerObjects.MiscData;

namespace MMR_Tracker_V3.TrackerObjects
{
    public abstract class CheckableLocation(InstanceData.TrackerInstance Parent)
    {
        private InstanceData.TrackerInstance _parent = Parent;
        public InstanceData.TrackerInstance GetParent() { return _parent; }
        public void SetParent(InstanceData.TrackerInstance parent) { _parent = parent; }

        public string ID { get; set; }
        public bool Available { get; set; } = false;
        public CheckState CheckState { get; set; } = CheckState.Unchecked;
        public bool Starred { get; set; }
        public bool Hidden { get; set; } = false;
        public int? Price { get; set; } = null;
        public char? Currency { get; set; } = null;
        public string DisplayName { get; set; }
        public RandomizedState RandomizedState { get; set; } = RandomizedState.Randomized;
        public InstanceData.ReferenceData referenceData { get; set; } = new InstanceData.ReferenceData();
        public virtual void GetPrice(out int outPrice, out char outCurrency)
        {
            outPrice = Price ?? -1;
            outCurrency = Currency ?? '$';
            return;
        }
        public virtual (int, char) GetPrice()
        {
            return (Price ?? -1, Currency ?? '$');
        }
        public virtual bool hasPrice()
        {
            var (Price, Cur) = GetPrice();
            return Price >= 0;
        }
        public virtual void SetPrice(int inPrice, char inCurrency = '\0')
        {
            if (inCurrency == '\0') { inCurrency = Currency ?? '$'; }
            Price = inPrice < 0 ? null : inPrice;
            Currency = inCurrency;
            return;
        }

        public abstract string GetName();
        public abstract LogicDictionaryData.DictionaryCheckableLocationEntry GetAbstractDictEntry();
        public abstract MiscData.CheckableLocationTypes LocationType();
    }
}

File: MMR Tracker V3/SpoilerLogHandling/Python/Class/CheckableLocation.py
from abc import ABC, abstractmethod
from MIscData import *

class CheckableLocation(ABC):
    def __init__(self, parent):
        self._parent = parent
        self.ID = None
        self.Available = False
        self.CheckState = MiscData.CheckState.Unchecked
        self.Starred = False
        self.Hidden = False
        self.Price = None
        self.Currency = None
        self.DisplayName = None
        self.RandomizedState = MiscData.RandomizedState.Randomized
        self.referenceData = InstanceData.ReferenceData()

    def GetParent(self):
        return self._parent

    def SetParent(self, parent):
        self._parent = parent

    @abstractmethod
    def GetName(self):
        pass

    @abstractmethod
    def GetAbstractDictEntry(self):
        pass

    @abstractmethod
    def LocationType(self):
        pass

    @abstractmethod
    def GetPrice(self):
        pass

    @abstractmethod
    def SetPrice(self, inPrice, inCurrency='\0'):
        pass

    def hasPrice(self):
        price, _ = self.GetPrice()
        return price is not None and price >= 0

    def GetPrice(self):
        return self.Price if self.Price is not None else -1, self.Currency if self.Currency is not None else '$'
